### ISSUE TITLE
Error in log when remove-node is done while writing (with consistent-topology-changes) - hints dir removal failed (as it was not empty)

### ISSUE THREAD

**Issue by nyh** [scylladb/scylladb#17023]
This is a spin-off of https://github.com/scylladb/scylladb/issues/17021.

The dtest `materialized_views_test.py`::`TestMaterializedViews`::`test_remove_node_during_mv_insert_3_nodes` (and also `test_remove_node_during_mv_insert_4_nodes` fail when consistent topology changes is enabled (tablets are not necessary to see this bug).

This is easily reproduced locally, with the dtest command line:
```
PYTHONPATH=../scylla-ccm pytest --tablets --cassandra-dir=$HOME/scylla/build/dev materialized_views_test.py::TestMaterializedViews::test_remove_node_during_mv_insert_3_nodes
```
(note that tablets aren't actually needed to reproduce this issue - it also reproduces with only consistent topology changes, without tablets)

This test starts a 3-node cluster and then "removes" the second node. 
The failure happens because the log of the first node the contains an unexpected error:
```
ERROR 2024-01-28 18:54:15,334 [shard 1:strm] hints_manager - Exception when draining 127.0.66.2: std::filesystem::__cxx11::filesystem_error (error system:39, filesystem error: remove failed: Directory not empty [/home/nyh/.dtest/dtest-8okv_vt3/test/node1/hints/1/127.0.66.2])
ERROR 2024-01-28 18:54:15,335 [shard 0:strm] hints_manager - Exception when draining 127.0.66.2: std::filesystem::__cxx11::filesystem_error (error system:39, filesystem error: remove failed: Directory not empty [/home/nyh/.dtest/dtest-8okv_vt3/test/node1/hints/0/127.0.66.2])
```

Looking a few lines up on what happens I see that:

1. Repair reports having done removenode_with_repair, with leaving node is 127.0.66.2.
2. raft_topology reports streaming completed and then various fences and stuff I don't understand.
3. gossip finally reports that "Removed endpoint 127.0.66.2".
4. At that points the hints manager is asked to drain 127.0.66.2 and complains that the directory is not empty.

I don't see any reason why the hints directory needs to be empty... The test continues the background load during this operation, and it's not surprising that in some instant, node 1 didn't know that node 2 was going down and saved something in a hint. This was a common race condition in the past, seen in issues like 
https://github.com/scylladb/scylladb/issues/10552
https://github.com/scylladb/scylla-enterprise/issues/2270

And was fixed in https://github.com/scylladb/scylladb/issues/5087 by PR https://github.com/scylladb/scylladb/pull/8387 by @piodul (Piotr, maybe you remember it?).
Apparently the fix is related to how the gossiper works, so it's not very surprising it broke when switching from gossiper to consistent_topology_changes.

To be honest, I don't understand why this race had to be fixed at all. Why do we care if a few hints get stored and then discarded? But perhaps there's a way to fix the fix to also work with consistent_topology_changes.

**Comment by bhalevy** [comment]
@kostja / @kbr-scylla who should own this?

**Comment by bhalevy** [comment]
@dawmd is this case being handled in https://github.com/scylladb/scylladb/pull/15567?

**Comment by bhalevy** [comment]
> @kostja / @kbr-scylla who should own this?

please assign owner

**Comment by kostja** [comment]
Generally @temichus owns all dtest failures with consistent-topology-changes. There is no point in filing issues without coordinating with him. This one could be a duplicate of one of the issues he's already reported.

**Comment by bhalevy** [comment]
> Generally @temichus owns all dtest failures with consistent-topology-changes. There is no point in filing issues without coordinating with him. This one could be a duplicate of one of the issues he's already reported.

We don't have such a coordination in our process
and for a goid reason, since it will create a bottleneck.
Whoever is opening an issue should do their best effort to locate existing issues and avoid unnecessary duplicates, but if they come to the conclusion that a new issue is warranted, how would they even know with whom to coordinate, and why would they wait fir the coordinator to vet their decision?
How would they communicate their findings to the coordinator? And how would we keep a paper-trail of the vetting decision?

The answer to the abovd is simply in a new github issue that should collect the data and initial analysis of the issue author, and then it's the job of the field-expert to go over the issue and add their pixie-dust to it.  It's possible that it is a duplicate of an existing issue, but the fact that it wasn't clear tobthe author means there's an apparent gap, and documenting the thought / analysis process that led to the decision about being a duplicate will help the next guy to see this issue in a self-service way, without having to coordinate with anyone.  Hence it's a more scalable process.

**Comment by kbr-scylla** [comment]
Looks to be the same as https://github.com/scylladb/scylladb/issues/17209

**Comment by dawmd** [comment]
> Looks to be the same as #17209

It is the same indeed.

The problem might be related to how topology changes are handled. When using legacy topology, there is no draining of the hint manager at all. We stop it as if the node were not leaving the cluster – I've just checked it, and the hint file *does* remain on disk in that case as well.

However, in the case of topology on Raft, we ask the hint manager to drain the hints towards the node. When draining, all of the hints should be sent no matter what, but it doesn't happen. We don't even attempt it as far as I understand the logs. I suspect the problem lies in `hint_sender::can_send()`, more specifically here:

```cpp
if (!_state.contains(state::ep_state_left_the_ring)) {
    const auto& tm = _shard_manager.local_db().get_token_metadata();
    const auto host_id = tm.get_host_id_if_known(end_point_key());
    _state.set_if<state::ep_state_left_the_ring>(!host_id || !tm.is_normal_token_owner(*host_id));
}
// send the hints out if the destination Node is part of the ring - we will send to all new replicas in this case
return _state.contains(state::ep_state_left_the_ring);
```

If I understand the semantics of draining correctly, when a node is leaving the cluster (and according to the logs when running the test using topology on Raft the target node *is* leaving), we should set the state to `state::ep_state_left_the_ring`. That would make sending hints proceed, but it doesn't happen, so at some point the function returns `false` and we stop sending hints.

I have other priorities now, so I'll only look at the issue thoroughly later, but I hope this made the picture a little bit less vague.

**Comment by nyh** [comment]
Another dtest that now fails on this unexpected error in the log is `materialized_views_test.py::TestMaterializedViews::test_mv_populating_from_existing_data_during_node_remove` (after more serious bugs that this test experienced were fixed)
(fails when running with "--tablets").

**Comment by piodul** [comment]
I've narrowed it down.

TLDR: Hinted handoff assumes that the node is already removed from token metadata when `drain_for` is called, but this is not true in general (at least in raft topology).

Modifying token metadata employs the RCU pattern, i.e. the current token metadata is copied, modified and - in the end - swapped in place of the previous token metadata. So, while the new token metadata is computed, the old one is still visible from other places in the code.

Topology on raft modifies token metadata in reaction to topology changes via `sync_raft_topology_nodes` - for example, in `storage_service::raft_ip_address_updater::on_endpoint_change`:

```c++
co_await _ss.mutate_token_metadata([this, hid](mutable_token_metadata_ptr t) {
    return _ss.sync_raft_topology_nodes(std::move(t), hid, {});
});
```

However, apart from modifying token metadata it also notifies implementors of `endpoint_lifecycle_subscriber` about nodes going up/down/joining/leaving. `storage_proxy` is one of them, and it calls `db::hints::manager::drain_for`:

```c++
void storage_proxy::on_leave_cluster(const gms::inet_address& endpoint) {
    // Discarding these futures is safe. They're awaited by db::hints::manager::stop().
    (void) _hints_manager.drain_for(endpoint);
    (void) _hints_for_views_manager.drain_for(endpoint);
}
```

The code that drains hints calls `can_send()` which - as @dawmd suspected in his comment above - doesn't set the `state::ep_state_left_the_ring` flag. This is because it still sees the old token metadata. Adding a 1 second sleep at the beginning of `drain_for` makes the tests pass, but this is a workaround.

I'm thinking about how to fix this properly.

**Comment by bhalevy** [comment]
> I've narrowed it down.
> 
> TLDR: Hinted handoff assumes that the node is already removed from token metadata when `drain_for` is called, but this is not true in general (at least in raft topology).
> 
> Modifying token metadata employs the RCU pattern, i.e. the current token metadata is copied, modified and - in the end - swapped in place of the previous token metadata. So, while the new token metadata is computed, the old one is still visible from other places in the code.
> 
> Topology on raft modifies token metadata in reaction to topology changes via `sync_raft_topology_nodes` - for example, in `storage_service::raft_ip_address_updater::on_endpoint_change`:
> 
> co_await _ss.mutate_token_metadata([this, hid](mutable_token_metadata_ptr t) {
>     return _ss.sync_raft_topology_nodes(std::move(t), hid, {});
> });
> However, apart from modifying token metadata it also notifies implementors of `endpoint_lifecycle_subscriber` about nodes going up/down/joining/leaving. `storage_proxy` is one of them, and it calls `db::hints::manager::drain_for`:
> 
> void storage_proxy::on_leave_cluster(const gms::inet_address& endpoint) {
>     // Discarding these futures is safe. They're awaited by db::hints::manager::stop().
>     (void) _hints_manager.drain_for(endpoint);
>     (void) _hints_for_views_manager.drain_for(endpoint);
> }
> The code that drains hints calls `can_send()` which - as [@dawmd](https://github.com/dawmd) suspected in his comment above - doesn't set the `state::ep_state_left_the_ring` flag. This is because it still sees the old token metadata. Adding a 1 second sleep at the beginning of `drain_for` makes the tests pass, but this is a workaround.
> 
> I'm thinking about how to fix this properly.

@piodul any progress on this?

**Comment by piodul** [comment]
> @piodul any progress on this?

Sorry for the radio silence. I discussed it with the cluster team since I posted my previous comment and I have an idea for a fix (postpone notifications after token metadata is updated, like it used to be before). I was preempted/distracted by other things but I'll give it more priority and try to implement it this asap (this week at least).

**Comment by mykaul** [comment]
I don't think it's a release blocker - though I'd be happy to see it fixed for 6.0.

**Comment by kbr-scylla** [comment]
It does cause some gating dtests to fail in raft-topology mode: test_hintedhandoff_removenode, test_replace_with_background_workload

**Comment by bhalevy** [comment]
This also manifest in dtest-gating tests with tablets.  Marking as release blocker based on the above comments.
